### PR TITLE
TOAZ-281 Implement Azure support in MonitorAtBoot.

### PR DESCRIPTION
**Pull request by Ghost-in-a-Jar**
This pr adds the implementation for MonitorAtBoot for azure apps. This included adding a codepath for createappv2message and deleteappv2message in the MonitorAtBoot.scala class. I also added some tests to the MonitorAtBootSpec file. 

Will test out the functionality in a bee before merging.

---
Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green (you can re-run automation tests with a comment saying `jenkins retest`
- [ ] Run the automation tests multiple times in parallel to weed out instability if applicable via a comment saying `multi-test`
- [ ] Squash and merge; Delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/MonitorAtBoot.scala [hunk 1]
```diff
@@ -351,6 +458,101 @@ class MonitorAtBoot[F[_]](publisherQueue: Queue[F, LeoPubsubMessage],
...
+      .flatMap(_.headOption)
+      .flatMap { r =>
+        if (useParent) r.resourceParentId.flatMap(_.split('/').lastOption)
+        else r.resourceName.orElse(r.resourceId.flatMap(_.split('/').lastOption))
+      }
+      .fold(
+        F.raiseError[String](
+          AppCreationException(s"${resourceType} resource with purpose ${purpose} not found in landing zone")
+        )
+      )(F.pure)
```
Comment (Ghost-in-a-Jar) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/MonitorAtBoot.scala]: These two fns are taken from app service, maybe they should be in a trait?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/MonitorAtBoot.scala [hunk 2]
```diff
@@ -157,61 +177,148 @@ class MonitorAtBoot[F[_]](publisherQueue: Queue[F, LeoPubsubMessage],
...
+                for {
+                  tokenOpt <- samDAO.getCachedArbitraryPetAccessToken(app.auditInfo.creator)
+                  userToken <- F.fromOption(
+                    tokenOpt,
+                    MonitorAtBootException(
+                      s"Unable to get pet token for ${app.auditInfo.creator} for app ${app.id} in Provisioning status",
+                      appContext.traceId
+                    )
+                  )
+                  workspaceId: WorkspaceId = app.workspaceId.getOrElse(
```
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/MonitorAtBoot.scala]: Can we use `F.fromOption` here (and in places where we need to throw exceptions)?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/MonitorAtBoot.scala [hunk 3]
```diff
@@ -157,61 +177,148 @@ class MonitorAtBoot[F[_]](publisherQueue: Queue[F, LeoPubsubMessage],
...
+                    org.http4s.Credentials.Token(AuthScheme.Bearer, userToken)
+                  )
+                  workspaceDescOpt <- wsmDao.getWorkspace(workspaceId, bearerAuth)
+                  workspaceDesc <- F.fromOption(workspaceDescOpt,
+                                                WorkspaceNotFoundException(workspaceId, appContext.traceId)
+                  )
+
+                  landingZoneResources <- getLandingZoneResources(workspaceDesc.spendProfile, bearerAuth)
+                  storageContainer <- wsmDao.getWorkspaceStorageContainer(workspaceId, bearerAuth)
+                  workspaceId = app.workspaceId.getOrElse(
```
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/MonitorAtBoot.scala]: Another `F.fromOption`

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/MonitorAtBoot.scala [hunk 4]
```diff
@@ -351,6 +458,101 @@ class MonitorAtBoot[F[_]](publisherQueue: Queue[F, LeoPubsubMessage],
         )
       case x => F.raiseError(MonitorAtBootException(s"Unexpected status for runtime ${runtime.id}: ${x}", traceId))
     }
+
+  private def getLandingZoneResources(billingProfileId: String, userToken: Authorization)(implicit
```
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/MonitorAtBoot.scala]: I don't really like duplicating this code -- can we put this method in a shared location?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/MonitorAtBoot.scala (hunk 1 above)
Comment (rtitle) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/MonitorAtBoot.scala]: Yes - let's not duplicate this code

**Comment by rtitle** [comment]
Logic LGTM, main comment is around sharing code between `MonitorAtBoot` and `AppServiceInterp`

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/MonitorAtBoot.scala (hunk 4 above)
Comment (Ghost-in-a-Jar) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/MonitorAtBoot.scala]: sounds good yeah

**Review by jordansafer — APPROVED** [review]
Looks good! I think we're starting to repeat the same pattern a lot for dealing with SAM/pet credentials for Azure, so a minor thought in the back of my head is that if we keep having this we should refactor a little down the road.

**Comment by jordansafer** [comment]
Looks like scalafmt strikes again

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/MonitorAtBoot.scala [hunk 5]
```diff
@@ -57,7 +62,9 @@ class MonitorAtBoot[F[_]](publisherQueue: Queue[F, LeoPubsubMessage],
       .parEvalMapUnordered(10) { case (a, n, c) =>
         for {
           now <- F.realTimeInstant
-          implicit0(traceId: Ask[F, TraceId]) = Ask.const[F, TraceId](TraceId(s"BootMonitor${now}"))
+          implicit0(traceId: Ask[F, AppContext]) = Ask.const[F, AppContext](
```
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/MonitorAtBoot.scala:65]: any reason why we need `AppContext` here?

File: http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/MonitorAtBootSpec.scala [hunk 6]
```diff
@@ -242,6 +335,22 @@ class MonitorAtBootSpec extends AnyFlatSpec with TestComponent with LeonardoTest
     res.unsafeRunSync()(cats.effect.unsafe.IORuntime.global)
   }
 
+  it should "ignore non-monitored apps in Azure" in isolatedDbTest {
```
Comment (Qi77Qi) [inline http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/MonitorAtBootSpec.scala:338]: what's non-monitored apps?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/MonitorAtBoot.scala (hunk 5 above)
Comment (Ghost-in-a-Jar) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/MonitorAtBoot.scala:65]: Yeah to string it through to handleApp, there is a downstream fn that requires an implicit AppContext

File: http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/MonitorAtBootSpec.scala (hunk 6 above)
Comment (Ghost-in-a-Jar) [reply http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/MonitorAtBootSpec.scala:338]: any app that doesnt have the status Prvisioning or Deleting when leo starts up

**Review by rtitle — APPROVED** [review]
I re-reviewed this and it looks good. Tests failed due to known Google groups issues. Going to merge this.

**Pull request merged**